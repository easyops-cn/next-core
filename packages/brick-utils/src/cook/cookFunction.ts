import { Node } from "@babel/types";
import { SimpleFunction } from "@next-core/brick-types";
import { raiseErrorFactory, walkFactory } from "./utils";
import { CookFunctionVisitor } from "./CookFunctionVisitor";
import {
  CookFunctionOptions,
  CookVisitorState,
  PrecookFunctionResult,
} from "./interfaces";
import { supply } from "./supply";

export function cookFunction<T extends SimpleFunction>(
  precooked: PrecookFunctionResult,
  { rules = {}, globalVariables }: CookFunctionOptions = {}
): T {
  const raiseError = raiseErrorFactory(precooked.source);
  const state: CookVisitorState<T> = {
    source: precooked.source,
    raiseError,
    scopeMapByNode: precooked.scopeMapByNode,
    scopeStack: [supply(precooked.attemptToVisitGlobals, globalVariables)],
    isRoot: true,
    rules,
  };
  walkFactory(CookFunctionVisitor, (node: Node) => {
    raiseError(SyntaxError, `Unsupported node type \`${node.type}\``, node);
  })(precooked.function, state);
  return state.cooked;
}
