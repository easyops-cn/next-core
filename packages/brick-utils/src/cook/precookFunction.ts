import { parseExpression } from "@babel/parser";
import { FunctionExpression, Node } from "@babel/types";
import {
  PrecookOptions,
  PrecookVisitorState,
  PrecookFunctionResult,
} from "./interfaces";
import { PrecookFunctionVisitor } from "./PrecookFunctionVisitor";
import { walkFactory } from "./utils";

export function precookFunction(
  source: string,
  options?: PrecookOptions
): PrecookFunctionResult {
  const func = parseExpression(source, {
    plugins: ["estree"],
    strictMode: true,
  }) as FunctionExpression;
  if (func.type !== "FunctionExpression") {
    throw new SyntaxError("Invalid function declaration");
  }
  const state: PrecookVisitorState = {
    scopeStack: [],
    attemptToVisitGlobals: new Set(),
    scopeMapByNode: new WeakMap(),
    isRoot: true,
  };
  walkFactory(
    options?.visitors
      ? { ...PrecookFunctionVisitor, ...options.visitors }
      : PrecookFunctionVisitor,
    (node: Node) => {
      // eslint-disable-next-line no-console
      console.warn(
        `Unsupported node type \`${node.type}\`: \`${source.substring(
          node.start,
          node.end
        )}\``
      );
    }
  )(func, state);
  return {
    source,
    function: func,
    attemptToVisitGlobals: state.attemptToVisitGlobals,
    scopeMapByNode: state.scopeMapByNode,
  };
}
