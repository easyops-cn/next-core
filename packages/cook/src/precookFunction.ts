import type { FunctionDeclaration } from "@babel/types";
import { parseAsEstree } from "./parse.js";
import { precook, PrecookOptions } from "./precook.js";

let ASTCache = new WeakMap<object, FunctionDeclaration>();

export interface PrecookFunctionOptions extends PrecookOptions {
  cacheKey?: object;
  typescript?: boolean;
}

export interface PrecookFunctionResult {
  function: FunctionDeclaration;
  attemptToVisitGlobals: Set<string>;
}

export function precookFunction(
  source: string,
  { typescript, cacheKey, ...restOptions }: PrecookFunctionOptions = {}
): PrecookFunctionResult {
  let func = cacheKey ? ASTCache.get(cacheKey) : undefined;
  if (!func) {
    func = parseAsEstree(source, { typescript });
    if (cacheKey) {
      ASTCache.set(cacheKey, func);
    }
  }
  const attemptToVisitGlobals = precook(func, restOptions);
  return {
    function: func,
    attemptToVisitGlobals,
  };
}

export function clearFunctionASTCache(): void {
  ASTCache = new WeakMap();
}
