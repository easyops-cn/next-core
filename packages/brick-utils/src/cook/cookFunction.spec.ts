import { cookFunction } from "./cookFunction";
import { precookFunction } from "./precookFunction";

jest.spyOn(console, "warn").mockImplementation(() => void 0);

describe("cookFunction", () => {
  it.each<
    [string, { source: string; cases: { args: unknown[]; result: unknown }[] }]
  >([
    [
      "lexical variables in block statement",
      {
        source: `
          function test(a) {
            {
              let a;
              a = 9;
            }
            return a;
          }
        `,
        cases: [
          {
            args: [1],
            result: 1,
          },
        ],
      },
    ],
    [
      "lexical variables in block statement of if",
      {
        source: `
          function test(a) {
            if (true) {
              let a;
              a = 9;
            }
            return a;
          }
        `,
        cases: [
          {
            args: [1],
            result: 1,
          },
        ],
      },
    ],
    [
      "lexical variables in block statement of switch",
      {
        source: `
          function test(a) {
            switch (true) {
              case true:
                let a;
                a = 9;
            }
            return a;
          }
        `,
        cases: [
          {
            args: [1],
            result: 1,
          },
        ],
      },
    ],
    [
      "update param variables in block statement",
      {
        source: `
          function test(a) {
            {
              a = 9;
            }
            return a;
          }
        `,
        cases: [
          {
            args: [1],
            result: 9,
          },
        ],
      },
    ],
    [
      "update lexical variables in block statement",
      {
        source: `
          function test(a) {
            let b = a;
            {
              b = 9;
            }
            return b;
          }
        `,
        cases: [
          {
            args: [1],
            result: 9,
          },
        ],
      },
    ],
    [
      "switch statements: general",
      {
        source: `
          function test(a) {
            let b;
            switch(a) {
              case 1:
                b = 'A';
                break;
              case 2:
                b = 'B';
                break;
              case 9:
                b = 'X';
                return 'Z';
              default:
                b = 'C';
            }
            return b;
          }
        `,
        cases: [
          {
            args: [1],
            result: "A",
          },
          {
            args: [2],
            result: "B",
          },
          {
            args: [3],
            result: "C",
          },
          {
            args: [9],
            result: "Z",
          },
        ],
      },
    ],
    [
      "switch statements: missing a break",
      {
        source: `
          function test(a) {
            let b = '';
            switch(a) {
              case 1:
                b += 'A';
              case 2:
                b += 'B';
                break;
              default:
                b = 'C';
            }
            return b;
          }
        `,
        cases: [
          {
            args: [1],
            result: "AB",
          },
          {
            args: [2],
            result: "B",
          },
          {
            args: [3],
            result: "C",
          },
        ],
      },
    ],
    [
      "switch statements: missing a break before default",
      {
        source: `
          function test(a) {
            let b = '';
            switch(a) {
              case 1:
                b += 'A';
                break;
              case 2:
                b += 'B';
              default:
                b += 'C';
            }
            return b;
          }
        `,
        cases: [
          {
            args: [1],
            result: "A",
          },
          {
            args: [2],
            result: "BC",
          },
          {
            args: [3],
            result: "C",
          },
        ],
      },
    ],
    [
      "if statements",
      {
        source: `
          function test(a) {
            if (a === 1) {
              return 'A';
            } else if (a === 2) {
              return 'B';
            } else if (a === 3) {
              return;
            } else {
              return 'C';
            }
          }
        `,
        cases: [
          {
            args: [1],
            result: "A",
          },
          {
            args: [2],
            result: "B",
          },
          {
            args: [3],
            result: undefined,
          },
          {
            args: [4],
            result: "C",
          },
        ],
      },
    ],
    [
      "object destructuring",
      {
        source: `
          function test(a, { r: d, ...e } = {}, ...f) {
            const { x: b = 9, ...c } = a;
            return {
              b,
              c,
              d,
              e,
              f
            };
          }
        `,
        cases: [
          {
            args: [
              {
                x: 1,
                y: 2,
                z: 3,
              },
              {
                r: 4,
                s: 5,
                t: 6,
              },
              7,
              8,
            ],
            result: {
              b: 1,
              c: {
                y: 2,
                z: 3,
              },
              d: 4,
              e: {
                s: 5,
                t: 6,
              },
              f: [7, 8],
            },
          },
          {
            args: [
              {
                y: 2,
                z: 3,
              },
            ],
            result: {
              b: 9,
              c: {
                y: 2,
                z: 3,
              },
              d: undefined,
              e: {},
              f: [],
            },
          },
        ],
      },
    ],
    [
      "array destructuring",
      {
        source: `
          function test(a, [d, ...e] = []) {
            const [ b = 9, ...c ] = a;
            return [ 0, ...c, b, d, e];
          }
        `,
        cases: [
          {
            args: [
              [1, 2, 3],
              [4, 5, 6],
            ],
            result: [0, 2, 3, 1, 4, [5, 6]],
          },
          {
            args: [[undefined, 2, 3]],
            result: [0, 2, 3, 9, undefined, []],
          },
        ],
      },
    ],
    [
      "recursive",
      {
        source: `
          function test(a) {
            return a + (a > 1 ? test(a - 1) : 0);
          }
        `,
        cases: [
          {
            args: [2],
            result: 3,
          },
          {
            args: [3],
            result: 6,
          },
        ],
      },
    ],
    [
      "var variables overload param variables",
      {
        source: `
          function test(a) {
            var a = 2;
            return a;
          }
        `,
        cases: [
          {
            args: [1],
            result: 2,
          },
        ],
      },
    ],
    [
      "functions overload params variables",
      {
        source: `
          function test(a) {
            function a() {}
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "function",
          },
        ],
      },
    ],
    [
      "var variables hoist in block statements",
      {
        source: `
          function test() {
            var b = typeof a;
            if (false) {
              var a;
            }
            return b;
          }
        `,
        cases: [
          {
            args: [],
            result: "undefined",
          },
        ],
      },
    ],
    [
      "functions after var variables initialized",
      {
        source: `
          function test(a) {
            var a = 'A';
            function a() {}
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "string",
          },
        ],
      },
    ],
    [
      "functions after var variables uninitialized",
      {
        source: `
          function test(a) {
            var a;
            function a() {}
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "function",
          },
        ],
      },
    ],
    [
      "functions before var variables initialized",
      {
        source: `
          function test(a) {
            function a() {}
            var a = 'A';
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "string",
          },
        ],
      },
    ],
    [
      "functions before var variables uninitialized",
      {
        source: `
          function test(a) {
            function a() {}
            var a;
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "function",
          },
        ],
      },
    ],
    [
      "functions before false conditional var variables initialized",
      {
        source: `
          function test(a) {
            function a() {}
            if (false) {
              var a = 'A';
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "function",
          },
        ],
      },
    ],
    [
      "functions before true conditional var variables initialized",
      {
        source: `
          function test(a) {
            function a() {}
            if (true) {
              var a = 'A';
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "string",
          },
        ],
      },
    ],
    [
      "functions before blocked var variables initialized",
      {
        source: `
          function test(a) {
            function a() {}
            {
              var a = 'A';
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "string",
          },
        ],
      },
    ],
    [
      "conditional functions after var variables uninitialized",
      {
        source: `
          function test(a) {
            var a;
            if (true) {
              function a() {}
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "number",
          },
        ],
      },
    ],
    [
      "conditional functions with param uninitialized",
      {
        source: `
          function test(a) {
            if (true) {
              function a() {}
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [],
            result: "undefined",
          },
        ],
      },
    ],
    [
      "blocked functions after var variables uninitialized",
      {
        source: `
          function test(a) {
            var a;
            {
              function a() {}
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "number",
          },
        ],
      },
    ],
    [
      "conditional functions after var variables uninitialized with no params",
      {
        source: `
          function test() {
            var a;
            if (true) {
              function a() {}
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [],
            result: "undefined",
          },
        ],
      },
    ],
    [
      "blocked functions after var variables uninitialized with no params",
      {
        source: `
          function test() {
            var a;
            {
              function a() {}
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [],
            result: "undefined",
          },
        ],
      },
    ],
    [
      "blocked functions",
      {
        source: `
          function test(a) {
            {
              function a() {}
            }
            return typeof a;
          }
        `,
        cases: [
          {
            args: [1],
            result: "number",
          },
        ],
      },
    ],
    [
      "hoisted functions",
      {
        source: `
          function test() {
            const t = a();
            function a() {
              return 1;
            }
            return t;
          }
        `,
        cases: [
          {
            args: [],
            result: 1,
          },
        ],
      },
    ],
    [
      "functions hoisting in block",
      {
        source: `
          function test() {
            const t = a();
            let r;
            if (true) {
              r = a();
              function a() {
                return 2;
              }
            }
            function a() {
              return 1;
            }
            return t + r;
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "functions hoisting in switch statement",
      {
        source: `
          function test() {
            const t = a();
            let r;
            switch (true) {
              case true:
                r = a();
              case false:
                function a() { return 2 }
            }
            function a() {
              return 1;
            }
            return t + r;
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "hoisted functions in function expressions",
      {
        source: `
          function test() {
            const f = function(){
              const t = a();
              let r;
              if (true) {
                r = a();
                function a() {
                  return 2;
                }
              }
              function a() {
                return 1;
              }
              return t + r;
            };
            return f();
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "hoisted functions in arrow functions",
      {
        source: `
          function test() {
            const f = () => {
              const t = a();
              let r;
              if (true) {
                r = a();
                function a() {
                  return 2;
                }
              }
              function a() {
                return 1;
              }
              return t + r;
            };
            return f();
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "for const ... of",
      {
        source: `
          function test() {
            let total = 0;
            for (const i of [1, 2]) {
              total += i;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "for var ... of",
      {
        source: `
          function test() {
            let total = 0;
            for (var i of [1, 2]) {
              total += i;
            }
            return total + i;
          }
        `,
        cases: [
          {
            args: [],
            result: 5,
          },
        ],
      },
    ],
    [
      "for let ... of and break",
      {
        source: `
          function test() {
            let total = 0;
            for (let i of [1, 2]) {
              total += i;
              if (total >= 1) {
                break;
                // Should never reach here.
                total += 10;
              }
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 1,
          },
        ],
      },
    ],
    [
      "for const ... of and continue",
      {
        source: `
          function test() {
            let total = 0;
            for (const i of [1, 2, 3]) {
              if (i === 2) {
                continue;
              }
              total += i;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 4,
          },
        ],
      },
    ],
    [
      "for let ... in",
      {
        source: `
          function test() {
            let total = '';
            for (let i in {a:1,b:2}) {
              total += i;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: "ab",
          },
        ],
      },
    ],
    [
      "for var ... in",
      {
        source: `
          function test() {
            let total = '';
            for (var i in {a:1,b:2}) {
              total += i;
            }
            return total + i;
          }
        `,
        cases: [
          {
            args: [],
            result: "abb",
          },
        ],
      },
    ],
    [
      "for const ... in and return",
      {
        source: `
          function test() {
            let total = '';
            for (let i in {a:1,b:2}) {
              total += i;
              if (total.length >= 1) {
                return 'oops: ' + total;
              }
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: "oops: a",
          },
        ],
      },
    ],
    [
      "for const ... in and continue",
      {
        source: `
          function test() {
            let total = '';
            for (const i in {a:1,b:2,c:3}) {
              if (i === 'b') {
                continue;
              }
              total += i;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: "ac",
          },
        ],
      },
    ],
    [
      "for let ...",
      {
        source: `
          function test() {
            let total = 0;
            const list = [1, 2];
            for (let i = 0; i < list.length; i += 1) {
              total += list[i];
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "for var ... and break",
      {
        source: `
          function test() {
            let total = 0;
            const list = [1, 2];
            for (var i = 0; i < list.length; i += 1) {
              total += list[i];
              if (total >= 1) {
                break;
              }
            }
            return total + i;
          }
        `,
        cases: [
          {
            args: [],
            result: 1,
          },
        ],
      },
    ],
    [
      "for const ... and continue",
      {
        source: `
          function test() {
            let total = 0;
            const list = [1, 2, 3];
            for (let i = 0; i < list.length; i += 1) {
              if (i === 1) {
                continue;
              }
              total += list[i];
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 4,
          },
        ],
      },
    ],
    [
      "for ... with no init nor test nor update",
      {
        source: `
          function test() {
            let total = 0;
            for (; ;) {
              total += 1;
              if (total >= 2) {
                break;
              }
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 2,
          },
        ],
      },
    ],
    [
      "nested for ...",
      {
        source: `
          function test() {
            let total = 0;
            const list = [1, 2];
            const object = {a: 3, b: 4};
            for (const i of list) {
              total += i;
              for (const k in object) {
                total += object[k];
              }
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 17,
          },
        ],
      },
    ],
    [
      "nested for ... and break inner",
      {
        source: `
          function test() {
            let total = 0;
            const list = [1, 2];
            const object = {a: 3, b: 4};
            for (const i of list) {
              total += i;
              for (const k in object) {
                total += object[k];
                break;
              }
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 9,
          },
        ],
      },
    ],
    [
      "nested for ... and break outer",
      {
        source: `
          function test() {
            let total = 0;
            const list = [1, 2];
            const object = {a: 3, b: 4};
            for (const i of list) {
              total += i;
              for (const k in object) {
                total += object[k];
              }
              break;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 8,
          },
        ],
      },
    ],
    [
      "nested for ... and return inner",
      {
        source: `
          function test() {
            let total = 0;
            const list = [1, 2];
            const object = {a: 3, b: 4};
            for (const i of list) {
              total += i;
              for (const k in object) {
                total += object[k];
                return "oops: " + total;
              }
              alert('yaks');
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: "oops: 4",
          },
        ],
      },
    ],
    [
      "while ...",
      {
        source: `
          function test() {
            let total = 0;
            while (total <= 2) {
              total += 1;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "while ... false",
      {
        source: `
          function test() {
            let total = 0;
            while (false) {
              total += 1;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 0,
          },
        ],
      },
    ],
    [
      "while ... and break",
      {
        source: `
          function test() {
            let total = 0;
            while (true) {
              total += 1;
              if (total >= 2) {
                break;
              }
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 2,
          },
        ],
      },
    ],
    [
      "while ... and continue",
      {
        source: `
          function test() {
            let total = 0;
            while (total < 2) {
              if (total >= 2) {
                continue;
              }
              total += 1;
            }
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 2,
          },
        ],
      },
    ],
    [
      "do ... while",
      {
        source: `
          function test() {
            let total = 0;
            do {
              total += 1;
            } while (total <= 2);
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 3,
          },
        ],
      },
    ],
    [
      "do ... while false",
      {
        source: `
          function test() {
            let total = 0;
            do {
              total += 1;
            } while (false);
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 1,
          },
        ],
      },
    ],
    [
      "do ... while and break",
      {
        source: `
          function test() {
            let total = 0;
            do {
              total += 1;
              if (total >= 2) {
                break;
              }
            } while (true);
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 2,
          },
        ],
      },
    ],
    [
      "do ... while and continue",
      {
        source: `
          function test() {
            let total = 0;
            do {
              if (total >= 2) {
                continue;
              }
              total += 1;
            } while (total < 2);
            return total;
          }
        `,
        cases: [
          {
            args: [],
            result: 2,
          },
        ],
      },
    ],
    [
      "empty statement",
      {
        source: `
          function test() {
            ;
          }
        `,
        cases: [
          {
            args: [],
            result: undefined,
          },
        ],
      },
    ],
    [
      "try ... catch ... finally",
      {
        source: `
          function test() {
            let a = 1, b, c;
            try {
              b = 'yep';
              a();
              b = 'nope';
            } catch (e) {
              a = e.toString();
            } finally {
              c = a + ':' + b;
            }
            return c;
          }
        `,
        cases: [
          {
            args: [],
            result: "TypeError: a is not a function:yep",
          },
        ],
      },
    ],
    [
      "throw and catch",
      {
        source: `
          function test() {
            let a = 'yes';
            try {
              throw 'oops';
            } catch (e) {
              a = 'Error: ' + e;
            }
            return a;
          }
        `,
        cases: [
          {
            args: [],
            result: "Error: oops",
          },
        ],
      },
    ],
    [
      "assignment",
      {
        source: `
          function test() {
            let [a, b, c, d, e, f, i] = [10, 20, 30, 40, 50, 60, 70];
            const g = { h: 70 };
            let r = a;
            let s = (b += 1);
            let t = (c -= 1);
            let u = (d *= 2);
            let v = (e /= 2);
            let w = (f %= 7);
            let x = (g.h += 2);
            let y = (i **= 2);
            return [[a, b, c, d, e, f, g.h, i], [r, s, t, u, v, w, x, y]];
          }
        `,
        cases: [
          {
            args: [],
            result: [
              [10, 21, 29, 80, 25, 4, 72, 4900],
              [10, 21, 29, 80, 25, 4, 72, 4900],
            ],
          },
        ],
      },
    ],
    [
      "update",
      {
        source: `
          function test() {
            let [a, b, c, d] = [10, 20, 30, 40];
            let r = a++;
            let s = ++b;
            let t = c--;
            let u = --d;
            return [ [a, b, c, d], [r, s, t, u] ];
          }
        `,
        cases: [
          {
            args: [],
            result: [
              [11, 21, 29, 39],
              [10, 21, 30, 39],
            ],
          },
        ],
      },
    ],
  ])("%s", (desc, { source, cases }) => {
    const func = cookFunction(precookFunction(source));
    for (const { args, result } of cases) {
      const equivalentFunc = new Function(`"use strict"; return (${source})`)();
      expect(equivalentFunc(...args)).toEqual(result);
      expect(func(...args)).toEqual(result);
    }
  });

  it("access global functions", () => {
    const source = `
      function test(name) {
        return FN.sayHello(name) + "!"
      }
    `;
    const func = cookFunction(precookFunction(source), {
      globalVariables: {
        FN: {
          sayHello(name: string) {
            return `Hello, ${name}`;
          },
        },
      },
    });
    expect(func("world")).toEqual("Hello, world!");
  });

  it("throw if restrict no war usage", () => {
    const source = `
      function test() {
        var a = 1;
        return a;
      }
    `;
    const equivalentFunc = new Function(`"use strict"; return (${source})`)();
    expect(() => equivalentFunc()).not.toThrowError();
    expect(() => {
      const func = cookFunction(precookFunction(source), {
        rules: {
          noVar: true,
        },
      });
      func();
    }).toThrowErrorMatchingSnapshot();
  });

  it.each<[desc: string, source: string]>([
    [
      "assign constants",
      `
        function test(){
          const a = 1;
          a = 2;
        }
      `,
    ],
    [
      "assign global functions",
      `
        function test(){
          test = 1;
        }
      `,
    ],
    [
      "assign function expressions",
      `
        function test(){
          (function f(){
            f = 1;
          })();
        }
      `,
    ],
    [
      "assign `for const ... of`",
      `
        function test(){
          for (const i of [1]) {
            i = 2;
          }
        }
      `,
    ],
    [
      "assign `for const ...`",
      `
        function test(){
          for (const i=0; i<2; i+=1) {
            i = 2;
          }
        }
      `,
    ],
    [
      "try without catch",
      `
        function test() {
          let a = 1, b, c;
          try {
            b = 'yep';
            a();
            b = 'nope';
          } finally {
            c = a + ':' + b;
          }
          return c;
        }
      `,
    ],
    [
      "assign member of nil",
      `
        function test() {
          let a;
          a.b = 1;
        }
      `,
    ],
    [
      "access before initialized",
      `
        function test() {
          let a = typeof b;
          let b;
          return a;
        }
      `,
    ],
  ])("%s should throw", (desc, source) => {
    const equivalentFunc = new Function(`"use strict"; return (${source})`)();
    expect(() => equivalentFunc()).toThrowError();
    expect(() => {
      const func = cookFunction(precookFunction(source));
      func();
    }).toThrowErrorMatchingSnapshot();
  });

  it.each<[desc: string, source: string]>([
    [
      "async functions",
      `
        async function test(){}
      `,
    ],
    [
      "generator functions",
      `
        function* test(){}
      `,
    ],
    [
      "bitwise and assignment",
      `
        function test() {
          let a = 1, b = 2;
          b &= a;
          return b;
        }
      `,
    ],
    [
      "labeled statement",
      `
        function test() {
          loop:
          for (let i = 0; i < 1; i++) {}
        }
      `,
    ],
  ])("%s should throw", (desc, source) => {
    const equivalentFunc = new Function(`"use strict"; return (${source})`)();
    expect(() => equivalentFunc()).not.toThrowError();
    expect(() => {
      const func = cookFunction(precookFunction(source));
      func();
    }).toThrowErrorMatchingSnapshot();
  });
});
