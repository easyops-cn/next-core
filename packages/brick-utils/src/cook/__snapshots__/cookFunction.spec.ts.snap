// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cookFunction access before initialized should throw 1`] = `"Cannot access 'b' before initialization"`;

exports[`cookFunction assign \`for const ... of\` should throw 1`] = `"Assignment to constant variable"`;

exports[`cookFunction assign \`for const ...\` should throw 1`] = `"Assignment to constant variable"`;

exports[`cookFunction assign constants should throw 1`] = `"Assignment to constant variable"`;

exports[`cookFunction assign function expressions should throw 1`] = `"Assignment to constant variable"`;

exports[`cookFunction assign global functions should throw 1`] = `"Assignment to constant variable"`;

exports[`cookFunction assign member of nil should throw 1`] = `"Cannot set property 'b' of undefined"`;

exports[`cookFunction async functions should throw 1`] = `"Async function is not allowed: \`async function test(){}\`"`;

exports[`cookFunction bitwise and assignment should throw 1`] = `"Unsupported assignment operator \`&=\`"`;

exports[`cookFunction generator functions should throw 1`] = `"Generator function is not allowed: \`function* test(){}\`"`;

exports[`cookFunction labeled statement should throw 1`] = `
"Unsupported node type \`LabeledStatement\`: \`loop:
          for (let i = 0; i < 1; i++) {}\`"
`;

exports[`cookFunction throw if restrict no war usage 1`] = `"Var declaration is not recommended, use \`let\` or \`const\` instead: \`var a = 1;\`"`;

exports[`cookFunction try without catch should throw 1`] = `"a is not a function"`;
